anesthetic/weighted_pandas.py

```diff
@@ -14,21 +14,38 @@ class _WeightedObject(object):
     def __init__(self, *args, **kwargs):
         weights = kwargs.pop('weights', None)
         super().__init__(*args, **kwargs)
-        self.weights = weights
+        if weights is None:
+            if 'weights' in self.index.names:
+                weights = self.weights
+            else:
+                weights = np.ones_like(self.index, int)
+        self._set_weights(weights)
 
     @property
     def weights(self):
         """Sample weights."""
-        if self.index.nlevels == 1:
-            return np.ones_like(self.index)
-        else:
-            return self.index.get_level_values('weights').to_numpy()
+        return self.index.get_level_values('weights').to_numpy()
 
     @weights.setter
     def weights(self, weights):
-        if weights is not None:
-            self.index = [self.index.get_level_values(0), weights]
-            self.index.set_names(['#', 'weights'], inplace=True)
+        self._set_weights(weights)
+
+    def _set_weights(self, weights):
+        self.index = [self.index.get_level_values(name)
+                      for name in self.index.names if name != 'weights'] \
+                          + [weights]
+        names = self.index.names[:-1] + ['weights']
+        self.index.set_names(names, inplace=True)
+
+    def reset_index(self, level=None, drop=False, inplace=False, *args):
+        weights = self.weights
+        answer = super().reset_index(level=level, drop=drop,
+                                     inplace=inplace, *args)
+        if inplace:
+            self._set_weights(weights)
+        else:
+            answer._set_weights(weights)
+            return answer
```

Comment:
Is `reset_index` behaving the way we'd expect it to? Try e.g.:
```python
print(ns.head())
ns.reset_index(level=0, inplace=True)
print(ns.head())
```

Reply:
I believe so. The key is to compares this to the results from `df = DataFrame(ns).copy()`, which I believe matches up. `reset_index` can be a bit counterintuitive at first.